mempool/clist_mempool_test.go

```diff
@@ -805,6 +785,211 @@ func TestMempoolNotifyTxsAvailable(t *testing.T) {
+	// last transactions.
+	mockClient.On("CheckTxAsync", mock.Anything, mock.Anything).Return(nil, nil).Times(6)
+	mockClient.On("CheckTxAsync", mock.Anything, mock.Anything).Return(nil, ErrCheckTxAsync{}).Times(2)
+
+	mp, cleanup := newMempoolWithAppMock(mockClient)
+	defer cleanup()
+	mp.EnableTxsAvailable()
+
+	// First send CheckTx requests for a few txs.
+	txs := []types.Tx{[]byte{0x01}, []byte{0x02}, []byte{0x03}, []byte{0x04}}
+	for _, tx := range txs {
+		_, err := mp.CheckTx(tx)
+		require.NoError(t, err)
+	}
+	require.Zero(t, mp.Size())
+
+	// Invoke CheckTx callbacks asynchronously to add txs to mempool.
+	for _, tx := range txs {
+		reqRes := newReqRes(tx, abci.CodeTypeOK, abci.CHECK_TX_TYPE_CHECK)
+		callback(reqRes.Request, reqRes.Response)
+	}
+	require.Len(t, txs, mp.Size())
+
+	// Check that recheck has not started yet.
+	require.True(t, mp.recheck.done())
+	require.Nil(t, mp.recheck.cursor)
+	require.Nil(t, mp.recheck.end)
+
+	// The first tx is valid, the third is invalid, and the request for the second and fourth tx
+	// were not sent, so the callback is not called.
+	mockClient.On("Flush", mock.Anything).Run(func(_ mock.Arguments) {
+		reqRes1 := newReqRes(txs[0], abci.CodeTypeOK, abci.CHECK_TX_TYPE_RECHECK)
+		callback(reqRes1.Request, reqRes1.Response)
+		reqRes2 := newReqRes(txs[2], 1, abci.CHECK_TX_TYPE_RECHECK)
+		callback(reqRes2.Request, reqRes2.Response)
+	}).Return(nil)
+
+	// mp.recheck.done() should be true only before and after calling recheckTxs.
+	mp.recheckTxs()
+	require.True(t, mp.recheck.done())
+	require.Nil(t, mp.recheck.cursor)
+	require.NotNil(t, mp.recheck.end)
+	require.Equal(t, mp.recheck.end, mp.txs.Back())
+	require.Equal(t, len(txs)-1, mp.Size()) // one invalid tx was removed
+	require.Equal(t, int32(2), mp.recheck.numPendingTxs.Load())
+
+	mockClient.AssertExpectations(t)
+}
+
+// This test used to cause a data race when rechecking (see https://github.com/cometbft/cometbft/issues/1827).
+func TestMempoolRecheckRace(t *testing.T) {
+	mp, cleanup := newMempoolWithAsyncConnection(t)
+	defer cleanup()
+
+	// Add a bunch of transactions to the mempool.
+	var err error
+	txs := newUniqueTxs(10)
+	for _, tx := range txs {
+		_, err = mp.CheckTx(tx)
+		require.NoError(t, err)
+	}
+
+	// Update one transaction to force rechecking the rest.
+	mp.Lock()
+	err = mp.FlushAppConn()
+	require.NoError(t, err)
+	err = mp.Update(1, txs[:1], abciResponses(1, abci.CodeTypeOK), nil, nil)
+	require.NoError(t, err)
+	mp.Unlock()
+
+	// Add again the same transaction that was updated. Checking this
+	// transaction and rechecking the others simultaneously should not result
+	// in a data race on the variable recheck.cursor.
+	_, err = mp.CheckTx(txs[:1][0])
+	require.Equal(t, err, ErrTxInCache)
+	require.Zero(t, mp.recheck.numPendingTxs.Load())
+}
+
+// Test adding txs while a concurrent routine reaps txs and updates the mempool, using an async ABCI
+// client.
+func TestMempoolConcurrentCheckTxAndUpdate(t *testing.T) {
```

Comment:
This test also passes with the original code (in `main`).

Reply:
This test have been very useful in debugging the code. It performs most mempool operations in the typical scenario where consensus updates the mempool while transactions are being added, concurrently and multiple times.